File: typing.cpp
// typing.cpp
#include <iostream>
#include <string>
#include <vector>
#include <random>
#include <chrono>
#include <fstream>
#include <map>
#include <algorithm>
#include <cctype>
#include <filesystem>

using namespace std;
using namespace std::chrono;
namespace fs = std::filesystem;

const string RESET = "\033[0m";
const string GREEN = "\033[92m";
const string RED = "\033[91m";
const string YELLOW = "\033[93m";
const string BLUE = "\033[94m";
const string BOLD = "\033[1m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

string getHomeDir() {
    const char* home = getenv("HOME");
    if (!home) home = getenv("USERPROFILE");
    return string(home);
}

map<string, vector<string>> TEXTS = {
    {"easy", {
        "кот и собака играют во дворе",
        "солнце светит ярко сегодня",
        "быстрый лис прыгает через забор"
    }},
    {"medium", {
        "программирование это искусство создавать алгоритмы для решения задач",
        "в мире информационных технологий важно постоянно учиться новому",
        "компьютерные науки охватывают множество направлений"
    }},
    {"hard", {
        "разработка программного обеспечения требует глубоких знаний и навыков работы в команде",
        "современные технологии изменяют мир и открывают новые возможности для человечества",
        "искусственный интеллект и машинное обучение становятся основой будущих инноваций"
    }}
};

class TypingTest {
public:
    string level;
    string text;
    string userInput;
    high_resolution_clock::time_point startTime, endTime;
    string statsFile;
    map<string, double> stats;

    TypingTest(string lvl) : level(lvl) {
        random_device rd;
        mt19937 gen(rd());
        auto& vec = TEXTS[level];
        uniform_int_distribution<> dis(0, vec.size()-1);
        text = vec[dis(gen)];
        statsFile = getHomeDir() + "/.typing_stats.json";
        loadStats();
    }

    void loadStats() {
        ifstream f(statsFile);
        if (!f) return;
        string content((istreambuf_iterator<char>(f)), istreambuf_iterator<char>());
        // Упрощённый парсинг JSON (без библиотеки)
        size_t pos = content.find("{");
        if (pos == string::npos) return;
        size_t end = content.rfind("}");
        if (end == string::npos) return;
        string json = content.substr(pos, end-pos+1);
        // Парсим пары "key":value
        size_t start = json.find("\"");
        while (start != string::npos) {
            size_t keyEnd = json.find("\"", start+1);
            if (keyEnd == string::npos) break;
            string key = json.substr(start+1, keyEnd-start-1);
            size_t valStart = json.find(":", keyEnd);
            if (valStart == string::npos) break;
            size_t valEnd = json.find(",", valStart+1);
            if (valEnd == string::npos) valEnd = json.find("}", valStart+1);
            if (valEnd == string::npos) break;
            string valStr = json.substr(valStart+1, valEnd-valStart-1);
            // Удаляем пробелы и кавычки
            valStr.erase(remove_if(valStr.begin(), valStr.end(), ::isspace), valStr.end());
            if (valStr.front() == '"') valStr.erase(0,1);
            if (valStr.back() == '"') valStr.pop_back();
            try {
                double val = stod(valStr);
                stats[key] = val;
            } catch (...) {}
            start = json.find("\"", valEnd);
        }
    }

    void saveStats() {
        ofstream f(statsFile);
        if (f) {
            f << "{";
            bool first = true;
            for (auto& kv : stats) {
                if (!first) f << ",";
                first = false;
                f << "\"" << kv.first << "\":" << kv.second;
            }
            f << "}";
        }
    }

    void displayText() {
        cout << colorize("\nНапечатайте следующий текст:", BOLD) << endl;
        cout << colorize(text, BLUE) << endl;
        cout << "\nНажмите Enter после ввода. Время начнётся с первой буквы." << endl;
        cout << "Готовы? Нажмите Enter...";
        cin.ignore();
        cout << "\nНачинайте печатать:" << endl;
    }

    void getUserInput() {
        startTime = high_resolution_clock::now();
        getline(cin, userInput);
        endTime = high_resolution_clock::now();
    }

    void calculateAndDisplay() {
        string original = text;
        string typed = userInput;
        int total = original.size();
        int typedLen = typed.size();
        int correct = 0;
        int errors = 0;
        for (int i=0; i<total; ++i) {
            if (i < typedLen && original[i] == typed[i]) {
                correct++;
            } else {
                errors++;
            }
        }
        if (typedLen < total) errors += total - typedLen;
        auto duration = duration_cast<milliseconds>(endTime - startTime).count();
        double seconds = duration / 1000.0;
        double minutes = seconds / 60.0;
        double wpm = (correct / 5.0) / minutes;
        double accuracy = (correct / (double)total) * 100.0;

        cout << colorize("\n📊 Результаты:", BOLD) << endl;
        cout << "  Скорость: " << wpm << " WPM" << endl;
        cout << "  Точность: " << accuracy << "%" << endl;
        cout << "  Правильных символов: " << correct << "/" << total << endl;
        cout << "  Ошибок: " << errors << endl;
        cout << "  Время: " << seconds << " сек" << endl;

        // Сохранение рекорда
        if (stats.find(level) == stats.end() || wpm > stats[level]) {
            stats[level] = wpm;
            saveStats();
            cout << colorize("🏆 Новый рекорд для этого уровня!", GREEN) << endl;
        } else {
            cout << "Лучший результат: " << stats[level] << " WPM" << endl;
        }
    }

    void run() {
        cout << colorize("⌨️  Тест скорости печати", BOLD) << endl;
        cout << "Уровень: " << level << endl;
        displayText();
        getUserInput();
        calculateAndDisplay();
    }
};

int main(int argc, char* argv[]) {
    string level = "medium";
    bool showStats = false, resetStats = false;
    string customText = "";
    for (int i=1; i<argc; ++i) {
        string arg = argv[i];
        if (arg == "easy" || arg == "medium" || arg == "hard") {
            level = arg;
        } else if (arg == "-t" || arg == "--text") {
            if (i+1 < argc) {
                customText = argv[++i];
            }
        } else if (arg == "-s" || arg == "--stats") {
            showStats = true;
        } else if (arg == "-r" || arg == "--reset") {
            resetStats = true;
        } else if (arg == "-h" || arg == "--help") {
            cout << "Usage: typing [easy|medium|hard] [-t text] [-s] [-r]" << endl;
            return 0;
        }
    }
    if (resetStats) {
        string f = getHomeDir() + "/.typing_stats.json";
        if (fs::exists(f)) fs::remove(f);
        cout << "Статистика сброшена." << endl;
        return 0;
    }
    if (showStats) {
        string f = getHomeDir() + "/.typing_stats.json";
        ifstream file(f);
        if (file) {
            string content((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
            // Простой вывод (для демонстрации)
            cout << colorize("📊 Статистика:", BOLD) << endl;
            // Парсим и выводим
            size_t pos = content.find("{");
            if (pos != string::npos) {
                size_t end = content.rfind("}");
                if (end != string::npos) {
                    string json = content.substr(pos+1, end-pos-1);
                    size_t start = 0;
                    while ((start = json.find("\"", start)) != string::npos) {
                        size_t keyEnd = json.find("\"", start+1);
                        if (keyEnd == string::npos) break;
                        string key = json.substr(start+1, keyEnd-start-1);
                        size_t valStart = json.find(":", keyEnd);
                        if (valStart == string::npos) break;
                        size_t valEnd = json.find(",", valStart+1);
                        if (valEnd == string::npos) valEnd = json.size();
                        string valStr = json.substr(valStart+1, valEnd-valStart-1);
                        valStr.erase(remove_if(valStr.begin(), valStr.end(), ::isspace), valStr.end());
                        if (valStr.front() == '"') valStr.erase(0,1);
                        if (valStr.back() == '"') valStr.pop_back();
                        try {
                            double val = stod(valStr);
                            cout << "  " << key << ": " << val << " WPM" << endl;
                        } catch (...) {}
                        start = valEnd;
                    }
                }
            }
        } else {
            cout << "Статистика пуста." << endl;
        }
        return 0;
    }
    if (!customText.empty()) {
        TypingTest test(level);
        test.text = customText;
        test.run();
    } else {
        TypingTest test(level);
        test.run();
    }
    return 0;
}

File: typing.go
// typing.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

var texts = map[string][]string{
	"easy": {
		"кот и собака играют во дворе",
		"солнце светит ярко сегодня",
		"быстрый лис прыгает через забор",
	},
	"medium": {
		"программирование это искусство создавать алгоритмы для решения задач",
		"в мире информационных технологий важно постоянно учиться новому",
		"компьютерные науки охватывают множество направлений",
	},
	"hard": {
		"разработка программного обеспечения требует глубоких знаний и навыков работы в команде",
		"современные технологии изменяют мир и открывают новые возможности для человечества",
		"искусственный интеллект и машинное обучение становятся основой будущих инноваций",
	},
}

type Stats map[string]float64

type TypingTest struct {
	level      string
	text       string
	userInput  string
	startTime  time.Time
	endTime    time.Time
	statsFile  string
	stats      Stats
}

func NewTypingTest(level string) *TypingTest {
	t := &TypingTest{level: level}
	t.statsFile = filepath.Join(os.Getenv("HOME"), ".typing_stats.json")
	t.loadStats()
	rand.Seed(time.Now().UnixNano())
	list := texts[level]
	t.text = list[rand.Intn(len(list))]
	return t
}

func (t *TypingTest) loadStats() {
	data, err := os.ReadFile(t.statsFile)
	if err != nil {
		t.stats = make(Stats)
		return
	}
	json.Unmarshal(data, &t.stats)
}

func (t *TypingTest) saveStats() {
	data, _ := json.MarshalIndent(t.stats, "", "  ")
	os.WriteFile(t.statsFile, data, 0644)
}

func (t *TypingTest) displayText() {
	fmt.Println(colorize("\nНапечатайте следующий текст:", bold))
	fmt.Println(colorize(t.text, blue))
	fmt.Println("\nНажмите Enter после ввода. Время начнётся с первой буквы.")
	fmt.Print("Готовы? Нажмите Enter...")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
	fmt.Println("\nНачинайте печатать:")
}

func (t *TypingTest) getUserInput() {
	t.startTime = time.Now()
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	t.userInput = strings.TrimRight(input, "\n")
	t.endTime = time.Now()
}

func (t *TypingTest) calculateAndDisplay() {
	original := t.text
	typed := t.userInput
	total := len(original)
	typedLen := len(typed)
	correct := 0
	errors := 0
	for i := 0; i < total; i++ {
		if i < typedLen && original[i] == typed[i] {
			correct++
		} else {
			errors++
		}
	}
	if typedLen < total {
		errors += total - typedLen
	}
	elapsed := t.endTime.Sub(t.startTime).Seconds()
	minutes := elapsed / 60.0
	wpm := (float64(correct) / 5.0) / minutes
	accuracy := (float64(correct) / float64(total)) * 100.0

	fmt.Println(colorize("\n📊 Результаты:", bold))
	fmt.Printf("  Скорость: %.1f WPM\n", wpm)
	fmt.Printf("  Точность: %.1f%%\n", accuracy)
	fmt.Printf("  Правильных символов: %d/%d\n", correct, total)
	fmt.Printf("  Ошибок: %d\n", errors)
	fmt.Printf("  Время: %.2f сек\n", elapsed)

	if val, ok := t.stats[t.level]; !ok || wpm > val {
		t.stats[t.level] = wpm
		t.saveStats()
		fmt.Println(colorize("🏆 Новый рекорд для этого уровня!", green))
	} else {
		fmt.Printf("Лучший результат: %.1f WPM\n", t.stats[t.level])
	}
}

func (t *TypingTest) run() {
	fmt.Println(colorize("⌨️  Тест скорости печати", bold))
	fmt.Printf("Уровень: %s\n", t.level)
	t.displayText()
	t.getUserInput()
	t.calculateAndDisplay()
}

func main() {
	level := "medium"
	showStats := false
	resetStats := false
	customText := ""
	args := os.Args[1:]
	for i := 0; i < len(args); i++ {
		arg := args[i]
		switch arg {
		case "easy", "medium", "hard":
			level = arg
		case "-t", "--text":
			if i+1 < len(args) {
				customText = args[i+1]
				i++
			}
		case "-s", "--stats":
			showStats = true
		case "-r", "--reset":
			resetStats = true
		case "-h", "--help":
			fmt.Println("Usage: typing [easy|medium|hard] [-t text] [-s] [-r]")
			return
		}
	}
	if resetStats {
		f := filepath.Join(os.Getenv("HOME"), ".typing_stats.json")
		os.Remove(f)
		fmt.Println("Статистика сброшена.")
		return
	}
	if showStats {
		f := filepath.Join(os.Getenv("HOME"), ".typing_stats.json")
		data, err := os.ReadFile(f)
		if err != nil {
			fmt.Println("Статистика пуста.")
			return
		}
		var stats Stats
		json.Unmarshal(data, &stats)
		fmt.Println(colorize("📊 Статистика:", bold))
		for lvl, wpm := range stats {
			fmt.Printf("  %s: %.1f WPM\n", lvl, wpm)
		}
		return
	}
	if customText != "" {
		test := NewTypingTest(level)
		test.text = customText
		test.run()
	} else {
		test := NewTypingTest(level)
		test.run()
	}
}
